Mobile - Add support visual test themes

geriux opened the pull request:
**Related PRs:**
- https://github.com/wordpress-mobile/gutenberg-mobile/pull/6187

## What?
This PR adds support to test themes in our E2E Visual testing environment.

## Why?
To prevent regressions related to theme colors, font size, and line height customization.

## How?
This is a follow-up of https://github.com/WordPress/gutenberg/pull/53434 that added the functionality to pass arguments during app launch.

We are adding `rawStyles` and `rawFeatures` in `initializeEditor` that then passes them through the `launchApp` util.

It also includes the native changes needed so the demo app supports these new arguments.

## Testing Instructions
CI checks should pass.

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast <!-- if applicable -->
N/A

Commit:
Android - Add support to pass rawStyles and rawFeatures during app launch

Commit:
iOS - Add support to pass rawStyles and rawFeatures during app launch

Commit:
Device tests - Editor Page: Pass rawStyles and rawFeatures through launchApp

Commit:
Update default values

fluiddot commented on packages/react-native-editor/ios/GutenbergDemo/GutenbergViewController.swift:
```diff
@@ -419,6 +430,26 @@ extension GutenbergViewController: GutenbergBridgeDataSource {
     }
 
     func gutenbergEditorSettings() -> GutenbergEditorSettings? {
+        guard ProcessInfo.processInfo.arguments.count >= 3 else {
+            return nil
+        }
+        
+        let initialProps = ProcessInfo.processInfo.arguments[2]
+
+        if let data = initialProps.data(using: .utf8) {
+            do {
+                if let json = try JSONSerialization.jsonObject(with: data, options: []) as? [String: Any] {
+                    let settings = EditorSettings()
+
+                    settings.rawStyles = json["rawStyles"] as? String
+                    settings.rawFeatures = json["rawFeatures"] as? String
+                    return settings
+                }
+            } catch {
+                print(error.localizedDescription)
+            }
+        }
+        
```
The initial props are processed in the function `getInitialPropsFromArgs`, so we could simplify this part by using that function. You can see an example in the function `gutenbergInitialContent`.

https://github.com/WordPress/gutenberg/blob/35573912f353c8125ff060f9af67df76cf890144/packages/react-native-editor/ios/GutenbergDemo/GutenbergViewController.swift#L390-L395

geriux replied:
Thanks for the suggestion! I will update it.

Commit:
Merge branch 'trunk' into rnmobile/add-support-visual-test-themes

Commit:
Update gutenbergEditorSettings logic

fluiddot reviewed (approved):
LGTM 🎊 !

Commit:
Merge branch 'trunk' into rnmobile/add-support-visual-test-themes

Commit:
Adds tapStatusBariOS util as well as adding a sleep timer in typeStringIos to wait for the auto-scroll animation to finish when typing

Commit:
Update getTitleElement to use tapStatusBariOS for iOS instead of swiping up for stability

Commit:
Update tapStatusBariOS to add a sleep timer to wait for the scroll animation to finish

geriux commented:
Hey @fluiddot 👋 

I wanted to mention that I included a few changes after your approval:

- [Adds tapStatusBariOS util as well as adding a sleep timer in typeStringIos to wait for the auto-scroll animation to finish when typing](https://github.com/WordPress/gutenberg/pull/54376/commits/998f7e5a9d03099a17e3cdf196ab4155022548a3)
- [Update getTitleElement to use tapStatusBariOS for iOS instead of swiping up for stability](https://github.com/WordPress/gutenberg/pull/54376/commits/bbeef2be8e88b108ec4c7cd5e6e55c8e3a35eda7)
- [Update tapStatusBariOS to add a sleep timer to wait for the scroll animation to finish](https://github.com/WordPress/gutenberg/pull/54376/commits/596e806356099fa96d96a9bcebcb267362a105ca)

I realized a flaky test was being reported due to not fully scrolling to the top on iOS so I updated the logic so it taps on the status bar to auto-scroll to the very top.

Besides that, I saw a constant issue when typing blocks on iOS, sometimes some tests would fail and I think it's related to the auto-scroll feature when typing to keep the content above the keyboard. So I added a sleep timer to avoid these issues.

Let me know if you have any doubts about those changes, thank you!

fluiddot commented on packages/react-native-editor/__device-tests__/pages/editor-page.js:
```diff
@@ -192,7 +193,11 @@ class EditorPage {
-			await swipeDown( this.driver );
+			if ( isAndroid() ) {
+				await swipeDown( this.driver );
+			} else {
+				await tapStatusBariOS( this.driver );
+			}
```
I didn't know that tapping in the iOS status bar makes the block list scroll up (TIL) 😮.

geriux replied:
I think it's the default for iOS and ScrollViews, I use it all the time 😃

fluiddot commented on packages/react-native-editor/__device-tests__/helpers/utils.js:
```diff
@@ -210,6 +210,9 @@ const typeStringIos = async ( driver, element, str, clear ) => {
 		await clearTextBox( driver, element );
 	}
 	await element.type( str );
+
+	// Wait for the list auto-scroll animation to finish
+	await driver.sleep( 3000 );
```
I presume that this change will increase the duration of E2E tests on iOS. However, unless we disable the scroll animation we can't do much else.

geriux replied:
Yeah, it will increase the run time a bit. I'll monitor that as well as if it reduces flakiness on some of the tests. We might be able to reduce this timer or just use it in some specific cases.
